### ISSUE TITLE
Support Conv3D

### ISSUE THREAD

**Issue by yeruoforever** [huggingface/candle#795]
3D convolutions are important too!

Like 2D images in nature, 3D Volume data is very common, such as medical CT and MRI data, where voxels are uniformly distributed in 3D space. 

The medical and health field relies heavily on computer vision methods that have migrated from the field of 2D natural image processing, such as classification, segmentation and detection of lesions and organs. These methods also rely on convolutional neural networks., even in the Transformer of the VIT family, 3D convolution is an elegant and efficient implementation of 3DPatchEmbed.

**Comment by janroden** [comment]
I'd love that too :-) 

Also for image registration (alignment) of 3D images (e.g. from MRI, CT, SPIM microscopy, ...), deep-learning based methods are getting better and more important (see e.g. [this recent paper on brain MRI registration](https://www.nature.com/articles/s41598-023-33781-0)). 

Having support for 3D images in Candle would enable us to implement such deep-learning based registration methods, and also "classical" optimization methods for 3D-image registration (see e.g. [this library using PyTorch for classical optimization-based image registration](https://github.com/airlab-unibas/airlab)).

This would bring the advantages of Candle to medical and biological image analysis, computer vision, etc., and could be used as a backbone to perform efficient (GPU) computation on 3D images in Rust in general.

Thanks a lot for this great library BTW!

**Comment by EricLBuehler** [comment]
Looks exciting, I wonder how complicated it would be to develop based on Conv2d?

**Comment by EricLBuehler** [comment]
@LaurentMazare, what are your thoughts on adding Conv3d to Candle - is it something that is planned to be added already? I would be happy to contribute an implementation.

**Comment by LaurentMazare** [comment]
I don't think it's planned yet so please go ahead if you want to give it a stab. What I would suggest is to start by implementing it as a custom op and together with a model that actually uses of it (not sure what is a good example of how these get used). Once you have some working model, we can decide between merging it within candle-core, or having it as a separate `candle-conv3d` crate based on the resulting code complexity + your experience implementing it. This way you will be able to iterate quickly on a separate repo + also try to the custom op layer quite a bit.

**Comment by EricLBuehler** [comment]
That sounds like a good idea - I think I will take a stab at it. Do you think I should implement a "toy", bespoke model trained on some 3d data or try to implement a more generic model (do you know of any that use conv3d)?

**Comment by LaurentMazare** [comment]
I feel that it would be much better if it's a real model. I wouldn't think that it's necessary to train it in candle but a model that has some already trained weights and we can run it on some data and check that it generates the same output as the python equivalent would be neat. Obviously it would be even better if it's a model that you find interesting/useful somehow.

**Comment by janroden** [comment]
Cool :) Thanks for wanting to give it a shot!

As an example, here's a relatively simple [Keras model for 3D image classification from CT scans](https://keras.io/examples/vision/3D_image_classification/). Note that this also needs `MaxPool3D` layers in addition to the `Conv3D` layers though. Also, unfortunately, it needs a bit of preprocessing of the images. I found [this crate for reading NIfTI images](https://github.com/Enet4/nifti-rs), but haven't used it yet.

**Comment by kingzcheung** [comment]
It seems that candle completely lacks operators related to x3d, such as BatchNorm3d, Conv3d, Dropout3d, MaxPool3d, ConvTranspose3d, etc.

https://github.com/zudi-lin/pytorch_connectomics/blob/1d8f16f7ab42d3395acffce8bf98897bb1bd5bc7/connectomics/model/backbone/repvgg.py

**Comment by SorenDreano** [comment]
This would definitely be a very important feature to have in such a framework. Many VLMs (including the Qwen-VL family) rely on Conv3d.

**Comment by ivarflakstad** [comment]
> This would definitely be a very important feature to have in such a framework. Many VLMs (including the Qwen-VL family) rely on Conv3d.

I agree we need dedicated 3d kernels, but note that we can (and do) use [this](https://github.com/huggingface/candle/blob/main/candle-transformers/src/models/qwen3_vl/conv3d_temporal_2.rs) right now.

In fact I think it would be fine to use this approach in a `Tensor::conv3d` method, as long as we document the expected performance and limitations of the approach.
